### PR TITLE
enable aggregate mem monitoring

### PR DESCRIPTION
**Pull request by vchiley**
# What does this PR do?

Enables memory monitor to aggregate stats across GPUs.
With dynamic graph execution, some GPUs may have more memory usage than others; this update allows the user to aggregate memory stats across GPUs.

# What issue(s) does this change relate to?

Discussed offline
Local testing via: https://github.com/mosaicml/llm-foundry-private/pull/153

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [x] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
  - [x] discussed offline
- [ ] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Review by mvpatel2000 — COMMENTED** [review]
Can you please add a unit test for this?

In particular I would write a distributed test where rank 0 creates a large tensor and then show rank 1 gets a higher mem usage

File: composer/callbacks/memory_monitor.py [hunk 1]
```diff
@@ -17,6 +19,37 @@
...
+    elif reduce_op == 'min':
+        op = distributed.ReduceOp.MIN
+    elif reduce_op == 'max':
+        op = distributed.ReduceOp.MAX
+    else:
+        raise ValueError(f'{reduce_op=} not supported.')
+
+    distributed.all_reduce(tensor_value, op=op)
+    if reduce_op == 'mean':
+        tensor_value = tensor_value / distributed.get_world_size()
```
Comment (mvpatel2000) [inline composer/callbacks/memory_monitor.py:48]:
```suggestion
    if reduce_op in ['mean', 'avg']:
        tensor_value = tensor_value / distributed.get_world_size()
```

File: composer/callbacks/memory_monitor.py [hunk 2]
```diff
@@ -101,6 +141,13 @@ def after_train_batch(self, state: State, logger: Logger):
             return
 
         memory_report = _get_memory_report(self.memory_keys)
+        if self.dist_aggregate_batch_interval is not None and state.timestamp.batch.value % self.dist_aggregate_batch_interval == 0:
+            dist_memory_report = {}
+            for (mem_stat, val) in memory_report.items():
+                dist_memory_report[mem_stat + '_avg'] = reduce_value(val, model_device, 'avg')
+                dist_memory_report[mem_stat + '_min'] = reduce_value(val, model_device, 'min')
+                dist_memory_report[mem_stat + '_max'] = reduce_value(val, model_device, 'max')
+            memory_report.update(dist_memory_report)
```
Comment (mvpatel2000) [inline composer/callbacks/memory_monitor.py:150]: Can you update docstring to include a comment saying these are provided if the interval is set?

**Review by mvpatel2000 — COMMENTED** [review]
LGTM. One small bug and a request for a comment

File: composer/callbacks/memory_monitor.py (hunk 2 above)
Comment (vchiley) [reply composer/callbacks/memory_monitor.py:150]: The `if` gating at [144](https://github.com/mosaicml/composer/pull/3042/files/14227f30492e283521d0da9ad8553a8eca717d45#diff-529d790253a002590be026ce3b6b59a345a08057be280fa3651ae055a7c495daR144) gives it away \s


yeah I'll add it.

Comment (vchiley) [reply composer/callbacks/memory_monitor.py]: https://github.com/mosaicml/composer/pull/3042/commits/278af0549522a4ebe99bbf4be4c1b81175638440

File: composer/callbacks/memory_monitor.py [hunk 3]
```diff
@@ -101,6 +141,14 @@ def after_train_batch(self, state: State, logger: Logger):
             return
 
         memory_report = _get_memory_report(self.memory_keys)
+        if self.dist_aggregate_batch_interval is not None and state.timestamp.batch.value % self.dist_aggregate_batch_interval == 0:
+            dist_memory_report = {}
+            for (mem_stat, val) in memory_report.items():
+                # only logged if dist_aggregate_batch_interval is set
```
Comment (mvpatel2000) [inline composer/callbacks/memory_monitor.py:147]:
```suggestion
```
Sorry I wasn't clear. The docstring for this class describes what values are logged in a table. We should add a comment in the class description that additional things are conditionally logged (no need to add to table)

Comment (vchiley) [reply composer/callbacks/memory_monitor.py]: [that](https://github.com/mosaicml/composer/pull/3042/commits/a2413de956a7670d75ff4eefcc8aaa385b7375bd) good

**Review by mvpatel2000 — APPROVED** [review]
🚀 🐈 🐒

**Pull request merged**